File: apps/csharp/design-pattern/tests/DesignPattern.Tests/Patterns/TemplateMethodTest.cs
using DesignPattern.Patterns.TemplateMethod;

namespace DesignPattern.Tests.Patterns;

public class TemplateMethodTest
{
    private readonly string[] _text = ["Hello", "World"];

    [Fact]
    public void HtmlReport_ContainsHtmlTags()
    {
        var report = new HtmlReport("Test", _text);
        var output = report.OutputReport();

        Assert.StartsWith("<html>", output);
        Assert.Contains("</html>", output);
    }

    [Fact]
    public void HtmlReport_ContainsTitle()
    {
        var report = new HtmlReport("Monthly Report", _text);
        var output = report.OutputReport();

        Assert.Contains("<title>Monthly Report</title>", output);
    }

    [Fact]
    public void HtmlReport_WrapsLinesInParagraphs()
    {
        var report = new HtmlReport("Test", _text);
        var output = report.OutputReport();

        Assert.Contains("<p>Hello</p>", output);
        Assert.Contains("<p>World</p>", output);
    }

    [Fact]
    public void PlainTextReport_ContainsFormattedTitle()
    {
        var report = new PlainTextReport("Test", _text);
        var output = report.OutputReport();

        Assert.Contains("***** Test *****", output);
    }

    [Fact]
    public void PlainTextReport_ContainsPlainLines()
    {
        var report = new PlainTextReport("Test", _text);
        var output = report.OutputReport();

        Assert.Contains("Hello\n", output);
        Assert.Contains("World\n", output);
    }

    [Fact]
    public void PlainTextReport_DoesNotContainHtmlTags()
    {
        var report = new PlainTextReport("Test", _text);
        var output = report.OutputReport();

        Assert.DoesNotContain("<html>", output);
        Assert.DoesNotContain("<p>", output);
    }
}

File: apps/scala/design-pattern/test/TemplateMethodTest.scala
package designpattern.templatemethod

class TemplateMethodSuite extends munit.FunSuite:

  test("HtmlReport が HTML 形式で出力する") {
    val report = HtmlReport()
    val output = report.outputReport()

    assert(output.contains("<html>"))
    assert(output.contains("<title>月次報告</title>"))
    assert(output.contains("<p>順調</p>"))
    assert(output.contains("<p>最高の調子</p>"))
    assert(output.contains("</html>"))
  }

  test("PlainTextReport がプレーンテキスト形式で出力する") {
    val report = PlainTextReport()
    val output = report.outputReport()

    assert(output.contains("**** 月次報告 ****"))
    assert(output.contains("順調"))
    assert(!output.contains("<html>"))
  }

  test("Report trait の outputLine は抽象メソッドのためコンパイル時に強制される") {
    // Scala では trait の抽象メソッドは実装しないとコンパイルエラー
    // JavaScript の throw Error とは異なり、型安全性で担保される
    assert(true)
  }
